Library architecture

konstantgr opened the pull request.

konstantgr committed:
implemented metrics of code complexity

metrics_utils.py file consists of functions which needed to evaluate code complexity metrics. Following metrics was made:

- SLOC - source lines of code

- comment_count  - number of comments in code

- Operation complexity - metric that consider different code structures with different weights

- classes_size = count of attributes + count of methods of a class

- NPAVG = all functions parameters / all functions count

- functions_count - all functions in cell (not only definitions)

- override_methods_count, new_methods_count - obvious

konstantgr committed:
First attempt of implementing library architecture

- connector/ folder consist of file connector, which contains class for getting data from localDb or Amazon and files that have already been used.

- processors/ folder contains file cell_processor.py, which contains class that inherits classes from files code_processor.py and md_processor.py. metrics_utils contains function for getting cyclomatic complexity and AST representation of source code. node_visitor.py is a file that contains a several classes for visiting nodes and getting statistics, such as different metrics or classes entities.

- notebook/ folder contains files for getting all previous knowledge in one data-structure and then write it in DB.

konstantgr committed:
add packanging managment using poetry

konstantgr committed:
fixed __init__.py files in all directories

konstantgr committed:
replace index.py with examples/app.py, added separate examples

TitovS commented on .gitignore:
```diff
@@ -1,11 +1,15 @@
-notebooks/
\ No newline at end of file
+.DS_Store
+commit.txt
```
strange

TitovS commented on connector/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from connector.connector import Connector
+from connector.db_structures import NotebookDb, CellDb, CodeCellDb, MdCellDb
+from connector.db_structures import create_db
```
from connector.connector import Connector
from connector.db_structures import NotebookDb, CellDb, CodeCellDb, MdCellDb
from connector.db_structures import create_db

```suggestion
from .connector import Connector
from .db_structures import NotebookDb, CellDb, CodeCellDb, MdCellDb
from .db_structures import create_db
```

TitovS commented on connector/connector.py:
```diff
@@ -0,0 +1,28 @@
+from sqlalchemy import create_engine
+from sqlalchemy.exc import OperationalError
+
+from . import get_data
```
better to do direct imports

`from .get_data import NotebookReaderDb, NotebookReaderAmazon`

TitovS commented on connector/connector.py:
```diff
@@ -0,0 +1,28 @@
+    def __init__(self, notebook_id, db_name):
+        self.data = None
+        self.engine = create_engine(f"sqlite:///{db_name}")
+        self.engine.dispose()
```
Why do we dispose engine just after init?

TitovS commented on connector/connector.py:
```diff
@@ -0,0 +1,28 @@
+        if isinstance(notebook_id, int):
+            try:
+                self.data = get_data.NotebookReaderDb(notebook_id, db_name)
+            except OperationalError as e:
```
We have already discussed - this is rare erorr can we place comment here why is it happening?

TitovS replied:
can this operation error be on `NotebookReaderDb` level?

TitovS commented on connector/get_data.py:
```diff
@@ -1,12 +1,11 @@
 from abc import ABC
 
-import db_structures
+from . import db_structures
```
better to do direct imports

TitovS commented on notebook/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from notebook.notebook import Notebook
```
same with . notation

TitovS commented on notebook/notebook.py:
```diff
@@ -0,0 +1,49 @@
+        connector = Connector(name, db_name)
+
+        self.engine = connector.engine
+        self.metadata = connector.data.get_notebook
```
can we rename get_notebook into get_metadata?

TitovS replied:
also maybe it will be direct methods of connector class (not properties?)

TitovS commented on notebook/write_to_db.py:
```diff
@@ -0,0 +1,63 @@
+    cell = db_structures.CellDb(notebook_id=notebook_id)
+    conn.add(cell)
+    conn.commit()
+    return cell.cell_id
```
Hm, maybe we could write all cells in one go - it will be a lot of transactions if will do one by one

TitovS commented on notebook/write_to_db.py:
```diff
@@ -0,0 +1,63 @@
+from connector import db_structures
+
+
+def write_notebook_to_db(conn, nb_metadata):
```
Maybe we can one one function which will write whole notebook - metadata and cells in one call (we dont use this functions separately)

TitovS commented on notebook/write_to_db.py:
```diff
@@ -0,0 +1,63 @@
+                           if not name.startswith('_')]
+        if key in (cell_attributes + ['content', 'metrics']):
+            if (key == 'content' and cell['type'] == 'markdown') \
+                    or (key == 'metrics' and cell['type'] == 'code'):  # TODO reconsider handling content
```
totaly agree with comment

TitovS replied:
need to flatten cell_attributes

TitovS commented on processors/__init__.py:
```diff
@@ -0,0 +1,6 @@
+import processors.metrics_utils
+from processors.cell_processor import CellProcessor
+from processors.code_processor import CodeProcessor
+from processors.md_processor import MdProcessor
```
import notation

TitovS commented on processors/cell_processor.py:
```diff
@@ -0,0 +1,20 @@
+            k: v
+            for k, v in tasks.items()
+            if (v and k in self.task_mapping.keys())
+        }:
```
Pretty heavy dict comprehension. Maybe on first step generate dict in on the second have for loop?
Also variable naming - k and v not very interpretable

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class CodeProcessor(CellProcessor):
+    task_mapping = {}
```
do we need creating this dict here? We are overwirting it a few strings later

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+            if (v and k in self.task_mapping.keys())
+        }:
+            self.cell[function] = self.task_mapping[function](self.cell)
+        return self.cell
```
You already have it is base class why do you override it? It is not abstract method so no need to re-define it.

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+        return 1 + max(
+            [self.depth_ast(child)
+             for child in ast.iter_child_nodes(cell_ast)],
+            default=0)
```
Can we do calculations in body of the function. Hard to read. Also magic number - better to name it.

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+        return len(cell['source'])
+
+    def get_imports(self, cell):
+        cell_ast = metrics_utils.get_ast(cell['source'])
```
It is second time you use gaet ast. Maybe get it on init? (if some of these functions are in config)

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+
+    def get_imports(self, cell):
+        cell_ast = metrics_utils.get_ast(cell['source'])
+        complexity_visitor = self.visitors[0]()
```
Can we have visitors in dict - [0] not very interpretable

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+            metrics['sloc'] = -1
+            metrics['comments_count'] = -1
+
+        ast_cell = metrics_utils.get_ast(source)  # ast.parse(source)
```
one more time

TitovS commented on processors/code_processor.py:
```diff
@@ -0,0 +1,90 @@
+            new_methods_count += len(params['new_methods'])
+            override_methods_count += len(params['override_methods'])
+
+        metrics['ccn'] = metrics_utils.get_cyclomatic_complexity(source)
+        metrics['operation_complexity'] = complexity_visitor.operation_complexity
+        metrics['classes_size'] = oop_visitor.classes_size
+        metrics['npavg'] = complexity_visitor.npavg
+        metrics['functions_count'] = len(complexity_visitor.functions)
+        metrics['override_methods_count'] = override_methods_count  # len(oop_visitor.override_methods)
+        metrics['new_methods_count'] = new_methods_count  # len(oop_visitor.new_methods)
```
For discussion:
maybe we split general metric in few function and separetes uses of different visitors in order to avoid overhead?

TitovS commented on processors/md_processor.py:
```diff
@@ -0,0 +1,68 @@
+
+        self.nlp = nlp
+
+    def process_cell(self, tasks) -> dict:
+        for function in {
+            k: v
+            for k, v in tasks.items()
+            if (v and k in self.task_mapping.keys())
+        }:
+            self.cell[function] = self.task_mapping[function](self.cell)
+        return self.cell
```
same as before - why override?

TitovS commented on processors/md_processor.py:
```diff
@@ -0,0 +1,68 @@
+        return self.cell
+
+    def get_cell_language(self, cell):
+        doc = self.nlp(cell['source'])
```
This is very expesive operation. Can we cache the result of nlp (same could be done with ast in previous processor)

TitovS commented on processors/metrics_utils.py:
```diff
@@ -0,0 +1,40 @@
+        return get_ast(code_string)
+
+
+def get_cyclomatic_complexity(cell_source):
```
why it is in utils? it seems that it part of code processor

TitovS commented on processors/metrics_utils.py:
```diff
@@ -0,0 +1,40 @@
+    if not refactored_cell:
+        return 0
+
+    try:
+        v = ComplexityVisitor.from_code(refactored_cell)
```
Oh recursive try-except. Can we find SyntaxError without executing from_code?

TitovS commented on processors/node_visitor.py:
```diff
@@ -0,0 +1,146 @@
+
+        if isinstance(node, (ast.AsyncFor, ast.While, ast.If,
+                             ast.With, ast.AsyncWith)):
+            self.operation_complexity += 0.5
```
Here and on on lines 21, 24, 28 (maybe somewhere else) - magic numbers. Maybe better to have object with weights for this numbers

konstantgr committed:
fixed defects

konstantgr committed:
reduced the number of commits in write_cells_to_db function from 2 * N_of_cells to 2

konstantgr committed:
added getting halstead vocavulary and refactored getting cyclomatic complexity

konstantgr committed:
Ready library architecture

- Fix collecting attributes and methods of classes
- Reafctored examples. There are 2 type of using library: collecting raw data and collecting prepared data. For raw data responds download_notebooks.py and get_notebooks.py for prepared data
